[Merged by Bors] - feat(analysis/normed_space/weak_dual): add the rest of Banach-Alaoglu theorem

kkytola opened the pull request:
The second of two parts to add the Banach-Alaoglu theorem about the compactness of the closed unit ball (and more generally polar sets of neighborhoods of the origin) of the dual of a normed space in the weak-star topology.

This second half is about the embedding of the weak dual of a normed space into a (big) product of the ground fields, and the required compactness statements from Tychonoff's theorem. In particular it contains the actual Banach-Alaoglu theorem.

Co-Authored-By: Yury Kudryashov <urkud@urkud.name>

---

- [x] depends on: #9827 [lemmas about bounds on linear maps over is_R_or_C]
- [x] depends on: #9836 [the first half of Banach-Alaoglu theorem, about polar sets]
- [x] depends on: #11112 


[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

kkytola committed:
feat(analysis/normed_space/operator_norm): add three lemmas on bounds of linear maps over is_R_or_C.

kkytola committed:
feat(analysis/normed_space/weak_dual): add Banach-Alaoglu theorem

kkytola committed:
Splitting to two halves: kept only the section about polar.

kkytola committed:
feat(analysis/normed_space/weak_dual): add the rest of Banach-Alaoglu theorem.

kkytola committed:
Merge branch 'master' into kkytola/banach_alaoglu_theorem

kkytola committed:
Import fix after merging master.

kkytola committed:
Merge branch 'kkytola/banach_alaoglu_theorem' into kkytola/banach_alaoglu_theorem_pt_2

kkytola committed:
Fixes after merging master + updates to docstrings.

kkytola committed:
Reverting to the master branch version of the file analysis.normed_space.bounded_linear_maps.

kkytola committed:
Reverting to the master branch version of the file analysis.normed_space.bounded_linear_maps here, too.

kkytola committed:
Apply suggestions from code review

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

kkytola committed:
Attempted starting changes, but need to get mathlib cache first.

kkytola committed:
Merge branch 'master' into kkytola/banach_alaoglu_theorem

kkytola committed:
Made the field arguments explicit and got rid of lots of at-symbols.

kkytola committed:
A big mess with closedness when trying to define polar sets in the normed dual instead.

kkytola committed:
Still a mess, but library search works again.

kkytola committed:
With an extra lemma, closedness of polar sets in the weak dual works.

kkytola committed:
Removing an unnecessary open and putting trivialities on single lines.

kkytola committed:
Moved the appropriate parts to the filed dual from the file weak_dual.

kkytola committed:
Correcting a docstring misprint due to the changed definition.

kkytola committed:
A tiny simplification I just learned.

kkytola committed:
Merge branch 'master' into kkytola/banach_alaoglu_theorem

kkytola committed:
Merge branch 'kkytola/banach_alaoglu_theorem' into kkytola/banach_alaoglu_theorem_pt_2

kkytola committed:
Updates after merges, including getting rid of at-symbols.

kkytola committed:
Also made the field an explicit argument in weak_dual.to_Pi.

kkytola committed:
Small fixes, not compiling yet.

kkytola committed:
Updates and fixes.

kkytola committed:
Sorry-free again.

kkytola committed:
Some suggested improvements implemented.

kkytola committed:
Follow naming convention: nhds.

kkytola committed:
Merging updates from branch kkytola/banach_alaoglu.

kkytola committed:
Fixes of things broken by changes.

kkytola committed:
Remove lemma finite_values_of_nhds_zero.

kkytola committed:
Merge branch 'kkytola/banach_alaoglu_theorem' into kkytola/banach_alaoglu_theorem_pt_2

sgouezel committed:
adapt to nondiscrete normed fields

kkytola committed:
Merge branch 'kkytola/banach_alaoglu_theorem' into kkytola/banach_alaoglu_theorem_pt_2

kkytola committed:
Generalized throughout with the help of Sebastien's improvement.

kkytola committed:
Modification to enable downstream generalizations based on Sebastien's improvement.

kkytola committed:
Merge with an easy resolved conflict.

kkytola committed:
An adaptation removing uninteresting lemmas.

kkytola committed:
Removed (essentially unnecessary) lemmas.

kkytola committed:
Removed a too special lemma and replaced by a more general one.

kkytola committed:
Apply suggestions from code review

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

kkytola committed:
Some fixes.

kkytola committed:
A better proof of eval_bounded_of_nhds_zero, as suggested.

kkytola committed:
A merge resolved.

kkytola committed:
A different way of talking about polar sets in weak_dual.

kkytola committed:
Simplifications, removing unnecessary lemmas.

kkytola committed:
Improvements from suggestions.

kkytola committed:
Merge branch 'kkytola/banach_alaoglu_theorem' into kkytola/banach_alaoglu_theorem_pt_2

kkytola committed:
Merged master and fixed merge conflicts.

kkytola committed:
Fixed the rest of merge conflicts.

kkytola committed:
Some updates.

kkytola committed:
Small updates to file level docstrings.

kkytola committed:
Updates to dosctrings and naming.

kkytola committed:
Clean-up within namespace and more updates to dosctrings and naming.

kkytola committed:
Update.

kkytola committed:
Update docstring: weak_dual.polar is an implementation detail.

sgouezel commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -58,9 +66,19 @@ Weak-* topology is defined generally in the file `topology.algebra.weak_dual_top
 
+For the proof of Banach-Alaoglu theorem, the weak dual of `E` is embedded in a product of copies
+of the ground field `𝕜` via `weak_dual.to_Pi : weak_dual 𝕜 E → Π (_ : E), 𝕜`. The fact that this
+is an embedding is  `weak_dual.to_Pi_embedding`.
```
```suggestion
is an embedding is `weak_dual.to_Pi_embedding`.
```

sgouezel commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +200,256 @@ begin
+
+/-- The (weak) dual `weak_dual 𝕜 E` of a normed space `E` consists of bounded linear
+functionals `E → 𝕜`. Such functionals can be interpreted as elements of the Cartesian
+product `Π (_ : E), 𝕜` via the function `weak_dual.to_Pi : weak_dual 𝕜 E → Π (_ : E), 𝕜`. -/
```
The cartesian product  `Π (_ : E), 𝕜` is by definition the very same thing as `E → 𝕜`. This means that many statements and docstrings below should be simplified accordingly.

kkytola replied:
Indeed, I now realize it is so much the same thing that the topology instance will be the same, and Tychonoff's theorem applies. This is probably what I wasn't confident about (although come think of it, I have used even the topology instance before).

Thank you! I will change to the better notation.

sgouezel commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +200,256 @@ begin
+/-- The function `weak_dual.to_Pi : weak_dual 𝕜 E → Π (_ : E), 𝕜` is an embedding. -/
+lemma weak_dual.to_Pi_embedding (𝕜 : Type*) [nondiscrete_normed_field 𝕜]
+  (E : Type*) [topological_space E] [add_comm_group E] [module 𝕜 E] :
+  embedding (λ (x' : weak_dual 𝕜 E), weak_dual.to_Pi 𝕜 E x') :=
```
```suggestion
  embedding (weak_dual.to_Pi 𝕜 E) :=
```

sgouezel commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +200,256 @@ begin
+  end, }
+
+lemma linear_of_mem_closure_range_apply
+  (f : (Π (_ : E), 𝕜)) (hf : f ∈ closure (range (@weak_dual.to_Pi 𝕜 _ E _ _ _))) (z : E) :
```
```suggestion
  (f : (Π (_ : E), 𝕜)) (hf : f ∈ closure (range (weak_dual.to_Pi 𝕜 E))) (z : E) :
```
should be enough.

sgouezel commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +200,256 @@ begin
+lemma continuous_of_mem_closure_polar
+  {s : set E} (s_nhd : s ∈ 𝓝 (0 : E)) (φ : (Π (_ : E), 𝕜))
+  (hφ : φ ∈ closure ((weak_dual.to_Pi 𝕜 E) '' (weak_dual.polar 𝕜 s))) :
+  @continuous E 𝕜 _ _ φ :=
```
```suggestion
  continuous φ :=
```
Again, no need for the `@` version, all the info is already there from the knowledge of phi.

sgouezel commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +200,256 @@ begin
+  apply mul_le_mul (hc x' polar_x') rfl.ge (norm_nonneg z) c_nn,
+end
+
+/-- The image under `weak_dual.to_Pi : weak_dual 𝕜 E → Π (_ : E), 𝕜` of a polar `polar s` of a
```
```suggestion
/-- The image under `weak_dual.to_Pi : weak_dual 𝕜 E → Π (_ : E), 𝕜` of a polar `polar 𝕜 s` of a
```
Same thing below.

sgouezel commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +200,256 @@ begin
+  simp only [one_div, inv_inv₀] at as_polar,
+  rw [←weak_dual.to_normed_dual.preimage_closed_ball, ←as_polar],
+  apply weak_dual.is_compact_polar (closed_ball_mem_nhds (0 : E) (inv_pos.mpr hr)),
+  exact finite_dimensional.proper_real 𝕜,
```
```suggestion
  exact weak_dual.is_compact_polar (closed_ball_mem_nhds (0 : E) (inv_pos.mpr hr)),
```
Sometimes `apply` has difficulties finding some instances, while `exact` solves them.

hrmacbeth commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +200,256 @@ begin
+are compact in the weak-star topology. -/
+theorem weak_dual.is_compact_closed_ball
+  {𝕜 : Type*} [is_R_or_C 𝕜] {E : Type*} [normed_group E] [normed_space 𝕜 E] (r : ℝ) (hr : 0 < r) :
+  is_compact {x' : weak_dual 𝕜 E | (∥ x'.to_normed_dual ∥ ≤ r)} :=
```
Perhaps, to make the geometric point clear,
```suggestion
  is_compact (metric.closed_ball (0 : normed_space.dual 𝕜 E) r : set (weak_dual 𝕜 E)) :=
```
This statement would be less unwieldy if you open the namespaces `metric` and `normed_space`.

kkytola replied:
In this form, this fails with "failed to synthesize type class instance for ... `pseudo_metric_space (weak_dual 𝕜 E)`", meaning that it is trying to create the metric ball in the wrong type synonym. Sure there are some ways to tell Lean to use the ball of `dual 𝕜 E` as a subset of `weak_dual 𝕜 E`, but my solutions for this have not always been pretty. I think something like `weak_dual.to_normed_dual ⁻¹' (metric.closed_ball (0 : normed_space.dual 𝕜 E) r)` is not terrible, but it still slightly obscures the geometric point.

kkytola replied:
Oh, but this allows to get rid of the clumsy rewrite-lemma `to_normed_dual.preimage_closed_ball` that I needed before!

Thank you!

hrmacbeth replied:
Oops, I should have tested it first!  One other way is
```lean
  is_compact (id (closed_ball 0 r : set (normed_space.dual 𝕜 E)) : set (weak_dual 𝕜 E))
```
but your current way also seems fine.

kkytola committed:
Apply suggestions from code review

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>
Co-authored-by: Heather Macbeth <25316162+hrmacbeth@users.noreply.github.com>

kkytola committed:
Changed Cartesian products to just functions.

kkytola committed:
Remove a lemma which became unnecessary.

kkytola committed:
Fix the noise in this noisy file.

kkytola committed:
A slightly different way of doing one lemma, but what is the best way?

kkytola committed:
Still don't know the best way, but this seems slightly less bad.

kkytola committed:
Stating the Banach-Alaoglu using id for the closed ball to get to the right type.

kkytola committed:
A few small golfs.

kkytola committed:
Corrected misformatted curly braces.

urkud commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +190,255 @@ begin
+form `{g | dist (f i) (g i) < ε}` with `ε > 0` and `i : ι` is a neighborhood of `f`. -/
+lemma _root_.mem_nhds_Pi' {ι : Type*} {X : ι → Type*} [∀ (i : ι), metric_space (X i)]
+  (f : (Π (j : ι), X j)) (i : ι) {ε : ℝ} (ε_pos : 0 < ε) :
+  {g : (Π (j : ι), X j) | dist (f i) (g i) < ε} ∈ 𝓝 f :=
```
This is just `continuous_at_apply i f (ball_mem_nhds (f i) ε_pos)` up to `dist_comm`. IMHO, this shouldn't be a separate lemma.

kkytola replied:
Thank you! I agree.

I just had not found a proof of less than ~7 lines beore. Also, t earlier versions the lemma `mem_nhds_Pi_normed_field` was used multiple times, so I had made it a standalone lemma.

I now removed these lemmas, and at the only remaining place where they are used, `norm_eval_le_of_mem_closure_norm_eval_le`, there is a 5 line subproof of a `have` statement. This reduces the number of lines overall (only one lemma becomes longer, but this is more than compensated by the removal of the dubious lemmas).

urkud commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +190,255 @@ begin
+form `{g | dist (f i) (g i) < ε}` with `ε > 0` and `i : ι` is a neighborhood of `f`. -/
+lemma _root_.mem_nhds_Pi_normed_field {ι : Type*} {K : ι → Type*} [∀ (i : ι), normed_group (K i)]
+  (f : (Π j, K j)) (i : ι) {ε : ℝ} (ε_pos : 0 < ε) :
+  {g : (Π j, K j) | ∥ f i - g i ∥ < ε} ∈ 𝓝 f :=
```
Same here (and if we add it anyway, then it should be named `normed_group.mem_nhds_pi` and live in `analysis.normed.group.basic`).

kkytola replied:
Thanks, I agree! I removed these.

urkud commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +190,255 @@ begin
+  (E : Type*) [topological_space E] [add_comm_group E] [module 𝕜 E] :
+  embedding (weak_dual.to_Pi 𝕜 E) :=
+{ induced := eq_of_nhds_eq_nhds (congr_fun rfl),
+  inj := by { intros φ₁ φ₂ h, ext z, exact congr_fun h z, }, }
```
Does `induced := rfl`, `inj := continuous_linear_map.coe_fn_injective` work?

kkytola replied:
It does! Thank you! Changed.

urkud commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +190,255 @@ begin
+respecting linear combinations. -/
+lemma linear_of_mem_closure_range'
+  (f : E → 𝕜) (hf : f ∈ closure (range (weak_dual.to_Pi 𝕜 E)))
+  (c₁ c₂ : 𝕜) (z₁ z₂ : E) : f (c₁ • z₁ + c₂ • z₂) = c₁ • f(z₁) + c₂ • f(z₂) :=
```
Note that we have `linear_map_of_tendsto`. We should add `linear_map.is_closed_range_coe` (and similarly for `(add_)monoid_hom`s) nearby. Since the range of coercion from `linear_map`s includes the range of coercion from `continuous_linear_map`s, this gives this lemma and next def.

kkytola replied:
Thank you! Very nice that `linear_map_of_tendsto` has appeared meanwhile!

I now added `linear_map.is_closed_range_coe`. I did not in fact use `linear_map_of_tendsto` in its proof; to me it seemed most natural to write the range of a coe from linear maps as an intersection of sets encoding the linearity constraints. It is not very elegant as it stands, but I think the overall strategy is natural, and I tend to think the intermediate steps could be useful elsewhere, too (say, constructing a linear map from the condition of respecting linear combinations).

urkud replied:
You're right that we should use `linear_map_of_mem_closure_range_coe` to define `linear_map_of_tendsto`, not vice versa. I implemented my solution in #11105. You're welcome to suggest changes there.

urkud commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +190,255 @@ begin
+  have hφ' : φ ∈ closure (range (weak_dual.to_Pi 𝕜 E)),
+  { apply mem_of_mem_of_subset hφ _,
+    apply closure_mono,
+    simp only [preimage_univ, preimage_range, subset_univ, image_subset_iff], },
```
Not tested:
```suggestion
  from closure_mono (image_subset_range _ _) hφ,
```

urkud commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +190,255 @@ begin
+    simp only [preimage_univ, preimage_range, subset_univ, image_subset_iff], },
+  set flin := linear_of_mem_closure_range φ hφ' with hflin,
+  suffices : continuous flin,
+  { assumption, },
```
If you prove that the range of `coe_fn` from `linear_map`s is closed, then you can `obtain ⟨φ, rfl⟩` that `φ` belongs to this range thus upgrading `φ` to a `linear_map`.

kkytola replied:
Sorry, I did not quite understand this comment.

I have proved `linear_map.is_closed_range_coe`, and I see how this improved and shortened other proofs, and I agree it is the right way generally, but I did not see an essential simplification here.

urkud commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +190,255 @@ begin
+  set θ := λ (ψ : E → 𝕜), ∥ ψ z ∥ with hθ,
+  have θ_cont : continuous θ,
+  { apply continuous.comp continuous_norm,
+    exact continuous_apply z, },
```
```suggestion
  from (continuous_apply z).norm,
```

kkytola replied:
I like this one-line-proof suggestion, and I don't see why it shouldn't work, but it doesn't. In this and variants, I get errors of the form:

```
type mismatch at application
  (continuous_apply z).norm
term
  continuous_apply z
has type
  continuous (λ (p : Π (i : E), ?m_1 i), p z)
but is expected to have type
  continuous ?m_5
```

I revert back to the two-line-proof for now.

urkud commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -182,3 +190,255 @@ begin
+
+/-- The image under `weak_dual.to_Pi : weak_dual 𝕜 E → (E → 𝕜)` of a polar `polar 𝕜 s` of a
+neighborhood `s` of the origin is a closed set. -/
+lemma image_polar_closed
```
```suggestion
lemma is_closed_image_polar
```

urkud reviewed (commented):
Thanks for working on this! I left some comments. Please fix the issues and switch the status back to `awaiting-review`.

kkytola committed:
Apply suggestions from code review

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

kkytola committed:
Removed dubious lemmas.

kkytola committed:
Fixed a too long line.

kkytola committed:
Golfed a proof according to a review suggestion.

git status

kkytola committed:
Merge branch 'master' into kkytola/banach_alaoglu_theorem_pt_2

kkytola committed:
Removed unused arguments and added docstring to a definition.

kkytola committed:
Add comments about the implementation of the new proposed strategy using linear_map.is_closed_range_coe.

kkytola committed:
A first attempt to refactor proofs using the suggested linear_map.is_closed_range_coe.

kkytola committed:
Added a further implementation comment.

kkytola committed:
Removed unnecessary special case lemmas.

kkytola committed:
Golf slightly.

urkud committed:
Merge branch 'master' of git://github.com/leanprover-community/mathlib into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
feat(topology/algebra): range of `coe_fn : (M₁ →ₛₗ[σ] M₂) → (M₁ → M₂)` is closed

Motivated by #9862

urkud committed:
Fix long line

urkud committed:
Merge branch 'YK-linear-of-closure' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Update src/topology/algebra/module.lean

urkud committed:
Snapshot

urkud committed:
Merge branch 'master' of git://github.com/leanprover-community/mathlib into YK-linear-of-closure

urkud committed:
Merge branch 'YK-linear-of-closure' of git://github.com/leanprover-community/mathlib into YK-linear-of-closure

urkud committed:
Add simps

urkud committed:
chore(topology/metric_space/basic): golf, add dot notation

* add `cauchy_seq.bounded_range`;
* golf `metric.bounded_range_of_tendsto`.

urkud committed:
Merge branch 'YK-linear-of-closure' into YK-clm-complete-split

urkud committed:
refactor(analysis/normed_space/operator_norm): split a proof

Split the proof of `continuous_linear_map.complete_space` into
reusable steps.

urkud committed:
Merge branch 'YK-clm-complete-split' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Snapshot

urkud committed:
feat(algebra/order/field): prove `a / a ≤ 1`

urkud committed:
Merge branch 'YK-div-self-le' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Revert

urkud committed:
Merge branch 'master' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Docs

urkud committed:
chore(topology/metric_space/basic): +1 version of Heine-Borel

* Mark `metric.is_compact_of_closed_bounded` as "Heine-Borel" theorem
  too.
* Add `metric.bounded.is_compact_closure`.

urkud committed:
Merge branch 'YK-heine-borel' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
feat(analysis/normed_space/dual): add lemmas, golf

* add `polar_univ`, `is_closed_polar`, `polar_gc`, `polar_Union`,
  `polar_union`, `polar_antitone`, `polar_zero`, `polar_closure`;

* extract `polar_ball_subset_closed_ball_div` and
  `closed_ball_inv_subset_polar_closed_ball` out of the proofs of
  `polar_closed_ball` and `polar_bounded_of_nhds_zero`;

* rename `polar_bounded_of_nhds_zero` to
  `bounded_polar_of_mem_nhds_zero`, use `metric.bounded`;

* use `r⁻¹` instead of `1 / r` in `polar_closed_ball`. This is the
  simp normal form (though we might want to change this in the future).

urkud committed:
Merge branch 'YK-polar-review' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Merge branch 'master' into YK-polar-review

urkud committed:
Merge branch 'YK-polar-review' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Docs

urkud committed:
Apply suggestions from code review

Co-authored-by: Patrick Massot <patrickmassot@free.fr>

urkud committed:
Merge branch 'master' into YK-linear-of-closure

urkud committed:
Don't use `@set.range`

urkud committed:
Merge branch 'YK-linear-of-closure' into YK-clm-complete-split

urkud committed:
Merge branch 'staging' of git://github.com/leanprover-community/mathlib into YK-linear-of-closure

urkud committed:
Merge branch 'YK-linear-of-closure' into YK-clm-complete-split

urkud committed:
Merge branch 'YK-polar-review' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Merge commit 'e92208aa6' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Merge remote-tracking branch 'origin/YK-linear-of-closure' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Merge commit '5bfd92457' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Merge commit 'be17b92a6' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Merge commit 'ea95523de' into YK-polar-review

urkud committed:
Merge branch 'YK-polar-review' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Merge branch 'master' into YK-linear-of-closure

urkud committed:
Merge branch 'YK-linear-of-closure' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Fix merge

urkud committed:
Use section vars

urkud committed:
Merge branch 'YK-clm-complete-split' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Snapshot

urkud committed:
chore(topology/algebra/weak_dual_topology): review

* Fix universes in `pi.has_continuous_smul`.
* Add `function.injective.embedding_induced` and
  `weak_dual.coe_fn_embedding`.
* Golf some proofs.
* Review `weak_dual.module` etc instances to ensure, e.g.,
  `module ℝ (weak_dual ℂ E)`.

urkud committed:
Merge branch 'YK-weak-top-review' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Snapshot

urkud committed:
Update

urkud committed:
Merge branch 'master' into kkytola/banach_alaoglu_theorem_pt_2

urkud committed:
Merge branch 'staging' of git://github.com/leanprover-community/mathlib into YK-clm-complete-split

urkud committed:
Merge branch 'YK-clm-complete-split' into kkytola/banach_alaoglu_theorem_pt_2

YaelDillies commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1261,6 +1261,74 @@ begin
+in a proper space. If `s` interpreted as a set in the space of maps `E → F` with topology of
+pointwise convergence is closed, then it is compact. -/
+lemma is_compact_image_coe_bounded_of_closed_image [proper_space F] {s : set (E' →SL[σ₁₂] F)}
+  (hb : bounded s) (hc : is_closed ((λ (f : E' →SL[σ₁₂] F) x, f x) '' s)) :
```
```suggestion
  (hb : bounded s) (hc : is_closed ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' s)) :
```

YaelDillies commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1261,6 +1261,74 @@ begin
+  is_compact ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' s) :=
+hc.closure_eq ▸ is_compact_closure_image_coe_bounded hb
+
+/-- If a set `s : set (E →SL[σ] F)` is bounded and is closed in the weak-* topology, then its image
```
It seems more useful to reformulate the Lean statement in human language than copy it. What do you think of this?
```suggestion
/-- If a set `s` of semilinear functions is bounded and is closed in the weak-* topology, then its image
```

YaelDillies commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1261,6 +1261,74 @@ begin
+at distance `≤ r` from `f₀ : E →SL[σ₁₂] F` is closed in the topology of pointwise convergence.
+This is one of the key steps in the proof of the **Banach-Alaoglu** theorem. -/
+lemma is_closed_image_coe_closed_ball (f₀ : E →SL[σ₁₂] F) (r : ℝ) :
+  is_closed ((λ (f : E →SL[σ₁₂] F) (x : E), f x) '' closed_ball f₀ r) :=
```
```suggestion
  is_closed ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' closed_ball f₀ r) :=
```

YaelDillies commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1261,6 +1261,74 @@ begin
+pointwise convergence. Other versions of this theorem can be found in
+`analysis.normed_space.weak_dual`. -/
+lemma is_compact_image_coe_closed_ball [proper_space F] (f₀ : E →SL[σ₁₂] F) (r : ℝ) :
+  is_compact ((λ (f : E →SL[σ₁₂] F) (x : E), f x) '' closed_ball f₀ r) :=
```
```suggestion
  is_compact ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' closed_ball f₀ r) :=
```

YaelDillies reviewed (commented):
Looks good!

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into kkytola/banach_alaoglu_theorem_pt_2

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into kkytola/banach_alaoglu_theorem_pt_2

YaelDillies committed:
bump

sgouezel commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -127,25 +128,90 @@ def continuous_linear_map_to_weak_dual : dual 𝕜 E →L[𝕜] weak_dual 𝕜 E
 
+variable {𝕜}
+
+/-- While coercion `coe_fn : weak_dual 𝕜 E → (E → 𝕜)` is not a closed map, it sends *bounded*
```
```suggestion
/-- While the coercion `coe_fn : weak_dual 𝕜 E → (E → 𝕜)` is not a closed map, it sends *bounded*
```

dupuisf commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1363,6 +1363,76 @@ begin
+-/
+lemma is_closed_image_coe_bounded_weak_closed {s : set (E' →SL[σ₁₂] F)} (hb : bounded s)
+  (hc : ∀ f, (⇑f : E' → F) ∈ closure ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' s) → f ∈ s) :
+  is_closed ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' s) :=
```
```suggestion
lemma is_closed_image_coe_of_bounded_of_weak_closed {s : set (E' →SL[σ₁₂] F)} (hb : bounded s)
  (hc : ∀ f, (⇑f : E' → F) ∈ closure ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' s) → f ∈ s) :
  is_closed ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' s) :=
```
I think it's a bit clearer this way.

dupuisf commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1363,6 +1363,76 @@ begin
+lemma is_compact_image_coe_bounded_weak_closed [proper_space F] {s : set (E' →SL[σ₁₂] F)}
+  (hb : bounded s)
+  (hc : ∀ f, (⇑f : E' → F) ∈ closure ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' s) → f ∈ s) :
+  is_compact ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' s) :=
```
```suggestion
lemma is_compact_image_coe_of_bounded_of_weak_closed [proper_space F] {s : set (E' →SL[σ₁₂] F)}
  (hb : bounded s)
  (hc : ∀ f, (⇑f : E' → F) ∈ closure ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' s) → f ∈ s) :
  is_compact ((coe_fn : (E' →SL[σ₁₂] F) → E' → F) '' s) :=
```

sgouezel commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1363,6 +1363,76 @@ begin
+/-- If a set `s` of semilinear functions is bounded and is closed in the weak-* topology, then its
+image under coercion to functions `E → F` is a closed set. We don't have a name for `E →SL[σ] F`
+with weak-* topology in `mathlib`, so we use an equivalent condition (see `is_closed_induced_iff'`).
+-/
```
Can you add a `TODO: reformulate this in terms of a type synonym with the right topology` in this lemma and the next ones?

sgouezel reviewed (commented):
LGTM, thanks!
bors d+

dupuisf commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -127,25 +128,90 @@ def continuous_linear_map_to_weak_dual : dual 𝕜 E →L[𝕜] weak_dual 𝕜 E
+lemma polar_def (s : set E) : polar 𝕜 s = {f : weak_dual 𝕜 E | ∀ x ∈ s, ∥f x∥ ≤ 1} := rfl
+
+/-- The polar `polar 𝕜 s` of a set `s : E` is a closed subset when the weak star topology
+is used, i.e., when `polar 𝕜 s` is interpreted as a subset of `weak_dual 𝕜 E`. -/
```
```suggestion
/-- The polar `polar 𝕜 s` of a set `s : E` is a closed subset when the weak star topology
is used. -/
```
I found the extra comment confusing, given that the statement is already about `weak_star.polar`.

dupuisf commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -127,25 +128,90 @@ def continuous_linear_map_to_weak_dual : dual 𝕜 E →L[𝕜] weak_dual 𝕜 E
+/-- While coercion `coe_fn : weak_dual 𝕜 E → (E → 𝕜)` is not a closed map, it sends *bounded*
+closed sets to closed sets. -/
+lemma is_closed_image_coe_bounded_closed {s : set (weak_dual 𝕜 E)}
+  (hb : bounded (dual.to_weak_dual ⁻¹' s)) (hc : is_closed s) :
```
```suggestion
lemma is_closed_image_coe_of_bounded_of_closed {s : set (weak_dual 𝕜 E)}
  (hb : bounded (dual.to_weak_dual ⁻¹' s)) (hc : is_closed s) :
```

dupuisf commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -127,25 +128,90 @@ def continuous_linear_map_to_weak_dual : dual 𝕜 E →L[𝕜] weak_dual 𝕜 E
+continuous_linear_map.is_closed_image_coe_bounded_weak_closed hb (is_closed_induced_iff'.1 hc)
+
+lemma is_compact_of_bounded_closed [proper_space 𝕜] {s : set (weak_dual 𝕜 E)}
+  (hb : bounded (dual.to_weak_dual ⁻¹' s)) (hc : is_closed s) :
```
```suggestion
lemma is_compact_of_bounded_of_closed [proper_space 𝕜] {s : set (weak_dual 𝕜 E)}
  (hb : bounded (dual.to_weak_dual ⁻¹' s)) (hc : is_closed s) :
```

dupuisf commented on src/analysis/normed_space/weak_dual.lean:
```diff
@@ -127,25 +128,90 @@ def continuous_linear_map_to_weak_dual : dual 𝕜 E →L[𝕜] weak_dual 𝕜 E
+/-- The image under `coe_fn : normed_space.dual 𝕜 E → (E → 𝕜)` of a polar `polar 𝕜 s` of a
+neighborhood `s` of the origin is a closed set. -/
+lemma _root_.normed_space.is_closed_image_polar {s : set E} (s_nhd : s ∈ 𝓝 (0 : E)) :
+  is_closed ((coe_fn : weak_dual 𝕜 E → E → 𝕜) '' polar 𝕜 s) :=
```
```suggestion
lemma _root_.normed_space.is_closed_image_polar_of_mem_nhds {s : set E} (s_nhd : s ∈ 𝓝 (0 : E)) :
  is_closed ((coe_fn : weak_dual 𝕜 E → E → 𝕜) '' polar 𝕜 s) :=
```

YaelDillies replied:
Also, we don't really use `normed_space` as a namespace.

j-loreaux replied:
we are using `normed_space.dual` as a namespace above, so I am changing it to that. (also note that we have things like [`normed_space.polar`](https://leanprover-community.github.io/mathlib_docs/analysis/normed_space/dual.html#normed_space.polar) already.

dupuisf reviewed (commented):
This is a really nice result to have in mathlib, thanks a lot!

I just had a few comments regarding lemma names, but otherwise LGTM.

j-loreaux committed:
generalize polar sets to `semi_normed_space`

j-loreaux committed:
use `semi_normed_group` throughout `analysis/normed_space/weak_dual`

j-loreaux committed:
fix docstrings

j-loreaux committed:
adjust some naming in `analysis/normed_space/weak_dual`

j-loreaux committed:
correct a lemma and rename it

j-loreaux committed:
add TODOs

j-loreaux committed:
adjust naming in `analysis/normed_space/operator_norm`

j-loreaux committed:
more naming fixes

j-loreaux committed:
Merge branch 'master' into kkytola/banach_alaoglu_theorem_pt_2
